File: D-MASH/client/backend/probe_primitives_v4.py
"""Route-scoped NCRH and bounded Probe hop budgets; not route authority."""
import hashlib
import hmac
import re
import secrets

MAX_HOPS = 15
DEFAULT_MIN_HOPS = 4


def _base(value):
    if not isinstance(value, bytes) or len(value) != 32:
        raise ValueError('BaseNCRH must be 32 secret bytes')
    return value


def route_ncrh(base_ncrh, route_id):
    if not isinstance(route_id, bytes) or len(route_id) != 32:
        raise ValueError('RouteID must be 32 decoded bytes')
    return hmac.new(_base(base_ncrh), b'D-MASH|NCRH|V4|ROUTE\0' + route_id, hashlib.sha256).hexdigest()


def extend_ncrh(base_ncrh, incoming):
    if not isinstance(incoming, str) or not re.fullmatch('[0-9a-f]{64}', incoming):
        raise ValueError('Invalid NCRH')
    return hmac.new(_base(base_ncrh), b'D-MASH|NCRH|V4|HOP\0' + bytes.fromhex(incoming), hashlib.sha256).hexdigest()


def sample_hop_ttl(minimum=DEFAULT_MIN_HOPS, maximum=MAX_HOPS):
    if type(minimum) is not int or type(maximum) is not int or not 1 <= minimum < maximum <= MAX_HOPS:
        raise ValueError('Invalid random hop range')
    return minimum + secrets.randbelow(maximum - minimum + 1)


def consume_hop(ttl):
    if type(ttl) is not int or not 1 <= ttl <= MAX_HOPS:
        raise ValueError('Invalid hop TTL')
    # Zero is a local stop sentinel, never an outgoing wire TTL.
    return ttl - 1

File: D-MASH PWA/not_messenger/js/probe_primitives_v4.js
'use strict';
(function(global){
 const MAX_HOPS=15,DEFAULT_MIN_HOPS=4;
 const bytes32=(value,name)=>{if(!(value instanceof Uint8Array)||value.length!==32)throw Error('Invalid '+name);return value;};
 const hex=value=>Array.from(value,b=>b.toString(16).padStart(2,'0')).join('');
 async function derive(base,domain,value){
  bytes32(base,'BaseNCRH');bytes32(value,'NCRH input');
  const prefix=new TextEncoder().encode('D-MASH|NCRH|V4|'+domain+'\0');
  const message=new Uint8Array(prefix.length+32);message.set(prefix);message.set(value,prefix.length);
  const key=await crypto.subtle.importKey('raw',base,{name:'HMAC',hash:'SHA-256'},false,['sign']);
  return hex(new Uint8Array(await crypto.subtle.sign('HMAC',key,message)));
 }
 function routeNcrh(base,routeId){return derive(base,'ROUTE',routeId);}
 function extendNcrh(base,incoming){
  if(typeof incoming!=='string'||!/^[0-9a-f]{64}$/.test(incoming))throw Error('Invalid NCRH');
  return derive(base,'HOP',Uint8Array.from(incoming.match(/../g),b=>parseInt(b,16)));
 }
 function sampleHopTtl(minimum=DEFAULT_MIN_HOPS,maximum=MAX_HOPS){
  if(!Number.isInteger(minimum)||!Number.isInteger(maximum)||minimum<1||minimum>=maximum||maximum>MAX_HOPS)throw Error('Invalid random hop range');
  const width=maximum-minimum+1,limit=Math.floor(0x100000000/width)*width;
  const sample=new Uint32Array(1);
  do{crypto.getRandomValues(sample);}while(sample[0]>=limit);
  return minimum+sample[0]%width;
 }
 function consumeHop(ttl){
  if(!Number.isInteger(ttl)||ttl<1||ttl>MAX_HOPS)throw Error('Invalid hop TTL');
  return ttl-1;
 }
 const api=Object.freeze({MAX_HOPS,DEFAULT_MIN_HOPS,routeNcrh,extendNcrh,sampleHopTtl,consumeHop});
 global.DmashProbePrimitivesV4=api;
 if(typeof module!=='undefined')module.exports=api;
})(typeof window!=='undefined'?window:globalThis);
